tests/cases/conformance/types/typeRelationships/typeInference/intraExpressionInferences.ts

```diff
@@ -176,3 +184,18 @@ example({
     fetch: (params: Params, foo) => 123,
     map: (number) => String(number)
 });
+
+// Repro from #45255
+
+declare const branch:
+  <T, U extends T>(_: { test: T, if: (t: T) => t is U, then: (u: U) => void }) => void
+
+declare const x: "a" | "b"
+
+branch({
+  test: x,
+  if: (t): t is "a" => t === "a",
+  then: u => {
+    let test1: "a" = u
+  }
+})

```

Comment:
This is very similar to some other cases here but this has an interesting return type annotation at `if`. Previously `u` in `then` was inferred to be of type `'a' | 'b'` but now it gets correctly narrowed down to `'a'`